src/main/kotlin/teaforge/platform/RoboRio/internal/TeaforgeImplementation.kt

```diff
@@ -171,12 +201,171 @@ fun <TMessage, TModel> processEffect(
+                Maybe.Some(it) to error
+            } ?: (Maybe.None to Maybe.Some(Error.SongNotLoaded))
+
+            val currentlyPlaying: Map<CanDeviceToken.MotorToken.TalonMotorToken, Orchestra> = orcAndError.first.map {
+                model.currentlyPlaying + (effect.motor to it)
+            }.valueOrDefault(model.currentlyPlaying)
+
+            model.copy(
+                loadedOrchestras = model.loadedOrchestras - effect.motor,
+                currentlyPlaying = currentlyPlaying
+            ) to Maybe.Some(effect.message(effect.motor, orcAndError.second))
+        }
+
+        is Effect.StopSong -> {
+
+            val error: Maybe<Error> = model.currentlyPlaying[effect.motor]?.let {
+                val result: StatusCode = it.stop()
+                it.close()
+                if (!result.isOK) Maybe.Some(Error.PhoenixError(result.name)) else Maybe.None
+            } ?: Maybe.Some(Error.SongNotPlaying)
+
+            model.copy(
+                currentlyPlaying = model.currentlyPlaying - effect.motor
+            ) to Maybe.Some(effect.message(effect.motor, error))
+        }
+
+        is Effect.SetCanMotorSpeed -> {
+            when (effect.motor) {
+                is CanDeviceToken.MotorToken.TalonMotorToken -> getCanDevice(effect.motor, model).set(effect.value)
+                is CanDeviceToken.MotorToken.NeoMotorToken -> getCanDevice(effect.motor, model).set(effect.value)
+            }
+            return model to Maybe.None
+        }
+
+        is Effect.ReadFile -> {
+            val result: Result<ByteArray, Error> = try {
+                val path: Path = Paths.get(effect.path)
+                val data: ByteArray = Files.readAllBytes(path)
+                Result.Success(data)
+            } catch (_: Exception) {
+                Result.Error(Error.InvalidFilename(effect.path))
+            }
+
+            model to Maybe.Some(effect.message(result))
+        }
+
+        is Effect.SetDioPort -> {
+            val errorMessage = effect.message(Result.Error(Error.AlreadyInitialized))
+            val successMessage = effect.message(Result.Success(effect.port))
+            val isInput: Boolean = model.dioInputs.containsKey(effect.port)
+            val alreadyInitialized: Boolean = model.dioOutputs.containsKey(effect.port)
+
+            if (isInput) model to Maybe.Some(errorMessage)
+
+            val newModel: RoboRioModel<TMessage, TModel> = if (!alreadyInitialized) {
+                model.copy(dioOutputs = model.dioOutputs + (effect.port to DigitalOutput(effect.port.id)))
+            } else {
+                model
+            }
+
+
+
+            return newModel.dioOutputs[effect.port]?.let {
+                val power: Boolean = when (effect.value) {
+                    DioPortState.LOW -> false
+                    DioPortState.HIGH -> true
+                }
+                it.set(power)
+                newModel to Maybe.Some(successMessage)
+            } ?: (newModel to Maybe.Some(errorMessage))
```

Comment:
This error state and the one on line 285 both return `Error.AlreadyInitialized` but seem to be caused by different situations. It would help debugging if we know exactly why this failed. in the first case the error is something like `DioPortAlreadyInitializedForInput` or `DioPortConfiguredForInput`.

The second error should never happen because of lines 287-291. If we get the value from the `model.dioOutputs` dictionary instead of just checking that it exists you won't have to fetch it again on line 295, either it will exist or you will have initialized a new value.